File: domain-expiration-approaching/python/check_domain_expiry.py
"""Check a domain's real expiration date over RDAP and alert when it crosses
a warning threshold. Renewal itself always happens at the registrar, since
this is a billing state, not a DNS zone record. DRY_RUN only reports until
turned off, and even then this script only sends alerts, it never renews
anything on your behalf.
"""
import os
import logging
from datetime import datetime, timezone

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("check_domain_expiry")

DNS_DOMAIN = os.environ.get("DNS_DOMAIN", "example.com")
CLOUDFLARE_API_TOKEN = os.environ.get("CLOUDFLARE_API_TOKEN", "")
CLOUDFLARE_ZONE_ID = os.environ.get("CLOUDFLARE_ZONE_ID", "")
DRY_RUN = os.environ.get("DRY_RUN", "true").lower() == "true"
WARN_THRESHOLDS_DAYS = [30, 14, 7, 1]


def days_until_expiry(expiration_iso: str, now_iso: str, warn_thresholds_days: list = None) -> dict:
    """
    Pure decision logic, no I/O.
    Input: expiration_iso (RDAP eventDate string, e.g. '2026-08-05T04:00:00Z'),
           now_iso (current time as ISO string, injected for testability),
           warn_thresholds_days (sorted descending list of alert thresholds).
    Output: {
        'days_remaining': int,
        'severity': 'ok' | 'warning' | 'critical' | 'expired',
        'triggered_threshold': int | None  # smallest threshold crossed, or None
    }
    Logic: parse both timestamps to UTC datetimes, compute
    days_remaining = floor((expiration - now).total_seconds() / 86400).
    If days_remaining < 0: severity='expired'.
    Else find the smallest threshold in warn_thresholds_days that is >= days_remaining;
    if none found, severity='ok', triggered_threshold=None;
    else severity='critical' if triggered_threshold <= 7 else 'warning'.
    """
    if warn_thresholds_days is None:
        warn_thresholds_days = [30, 14, 7, 1]

    expiration = datetime.fromisoformat(expiration_iso.replace("Z", "+00:00")).astimezone(timezone.utc)
    now = datetime.fromisoformat(now_iso.replace("Z", "+00:00")).astimezone(timezone.utc)
    days_remaining = int((expiration - now).total_seconds() // 86400)

    if days_remaining < 0:
        return {"days_remaining": days_remaining, "severity": "expired", "triggered_threshold": None}

    candidates = [t for t in warn_thresholds_days if t >= days_remaining]
    triggered_threshold = min(candidates) if candidates else None

    if triggered_threshold is None:
        severity = "ok"
    elif triggered_threshold <= 7:
        severity = "critical"
    else:
        severity = "warning"

    return {
        "days_remaining": days_remaining,
        "severity": severity,
        "triggered_threshold": triggered_threshold,
    }


def fetch_rdap_expiration(domain: str) -> str:
    """Query RDAP for a domain and return the expiration eventDate string."""
    import requests

    r = requests.get(f"https://rdap.org/domain/{domain}", timeout=30)
    r.raise_for_status()
    data = r.json()
    for event in data.get("events", []):
        if event.get("eventAction") == "expiration":
            return event["eventDate"]
    raise ValueError(f"No expiration event found in RDAP response for {domain}")


def fetch_rdap_status(domain: str) -> list:
    """Query RDAP for a domain and return its status list."""
    import requests

    r = requests.get(f"https://rdap.org/domain/{domain}", timeout=30)
    r.raise_for_status()
    return r.json().get("status", [])


def send_alert(domain: str, result: dict, statuses: list) -> None:
    """Send an alert. Replace with email, Slack, or a webhook call in production."""
    if DRY_RUN:
        log.info(
            "[dry run] would alert: %s is %s, %d day(s) remaining, statuses=%s",
            domain, result["severity"], result["days_remaining"], statuses,
        )
        return
    log.warning(
        "ALERT: %s is %s, %d day(s) remaining, statuses=%s. Renew at the registrar, "
        "this cannot be fixed through the DNS provider API.",
        domain, result["severity"], result["days_remaining"], statuses,
    )


def run():
    log.info("Checking domain expiration for %s (DRY_RUN=%s)", DNS_DOMAIN, DRY_RUN)

    expiration_iso = fetch_rdap_expiration(DNS_DOMAIN)
    statuses = fetch_rdap_status(DNS_DOMAIN)
    now_iso = datetime.now(timezone.utc).isoformat()

    result = days_until_expiry(expiration_iso, now_iso, WARN_THRESHOLDS_DAYS)
    log.info(
        "%s expires %s: %d day(s) remaining, severity=%s",
        DNS_DOMAIN, expiration_iso, result["days_remaining"], result["severity"],
    )

    grace_flags = {"pendingdelete", "redemptionperiod", "autorenewperiod"}
    lowered_statuses = [s.lower() for s in statuses]
    if any(flag in lowered_statuses for flag in grace_flags):
        log.warning("%s is already in a grace or pending-delete state: %s", DNS_DOMAIN, statuses)
        send_alert(DNS_DOMAIN, result, statuses)
        return

    if result["severity"] in ("warning", "critical", "expired"):
        send_alert(DNS_DOMAIN, result, statuses)
    else:
        log.info("OK: %s has plenty of runway left before renewal is needed.", DNS_DOMAIN)


if __name__ == "__main__":
    run()

File: domain-expiration-approaching/node/check-domain-expiry.js
/**
 * Check a domain's real expiration date over RDAP and alert when it crosses
 * a warning threshold. Renewal itself always happens at the registrar, since
 * this is a billing state, not a DNS zone record. DRY_RUN only reports until
 * turned off, and even then this script only sends alerts, it never renews
 * anything on your behalf.
 */
import { pathToFileURL } from "node:url";

const DNS_DOMAIN = process.env.DNS_DOMAIN || "example.com";
const CLOUDFLARE_API_TOKEN = process.env.CLOUDFLARE_API_TOKEN || "";
const CLOUDFLARE_ZONE_ID = process.env.CLOUDFLARE_ZONE_ID || "";
const DRY_RUN = (process.env.DRY_RUN || "true").toLowerCase() === "true";
const WARN_THRESHOLDS_DAYS = [30, 14, 7, 1];

/**
 * Pure decision logic, no I/O.
 * Input: expirationIso (RDAP eventDate string, e.g. '2026-08-05T04:00:00Z'),
 *        nowIso (current time as ISO string, injected for testability),
 *        warnThresholdsDays (sorted descending array of alert thresholds).
 * Output: {
 *   daysRemaining: number,
 *   severity: 'ok' | 'warning' | 'critical' | 'expired',
 *   triggeredThreshold: number | null
 * }
 * Logic: parse both timestamps, compute
 * daysRemaining = floor((expiration - now) / 86400000).
 * If daysRemaining < 0: severity = 'expired'.
 * Else find the smallest threshold in warnThresholdsDays that is >= daysRemaining;
 * if none found, severity = 'ok', triggeredThreshold = null;
 * else severity = 'critical' if triggeredThreshold <= 7 else 'warning'.
 */
export function daysUntilExpiry(expirationIso, nowIso, warnThresholdsDays = [30, 14, 7, 1]) {
  const expiration = new Date(expirationIso);
  const now = new Date(nowIso);
  const daysRemaining = Math.floor((expiration.getTime() - now.getTime()) / 86400000);

  if (daysRemaining < 0) {
    return { daysRemaining, severity: "expired", triggeredThreshold: null };
  }

  const candidates = warnThresholdsDays.filter((t) => t >= daysRemaining);
  const triggeredThreshold = candidates.length > 0 ? Math.min(...candidates) : null;

  let severity;
  if (triggeredThreshold === null) {
    severity = "ok";
  } else if (triggeredThreshold <= 7) {
    severity = "critical";
  } else {
    severity = "warning";
  }

  return { daysRemaining, severity, triggeredThreshold };
}

/** Query RDAP for a domain and return the expiration eventDate string. */
async function fetchRdapExpiration(domain) {
  const res = await fetch(`https://rdap.org/domain/${domain}`);
  if (!res.ok) throw new Error(`RDAP returned ${res.status} for ${domain}`);
  const data = await res.json();
  const event = (data.events || []).find((e) => e.eventAction === "expiration");
  if (!event) throw new Error(`No expiration event found in RDAP response for ${domain}`);
  return event.eventDate;
}

/** Query RDAP for a domain and return its status list. */
async function fetchRdapStatus(domain) {
  const res = await fetch(`https://rdap.org/domain/${domain}`);
  if (!res.ok) throw new Error(`RDAP returned ${res.status} for ${domain}`);
  const data = await res.json();
  return data.status || [];
}

/** Send an alert. Replace with email, Slack, or a webhook call in production. */
function sendAlert(domain, result, statuses) {
  if (DRY_RUN) {
    console.log(
      `[dry run] would alert: ${domain} is ${result.severity}, ${result.daysRemaining} day(s) remaining, statuses=${JSON.stringify(statuses)}`
    );
    return;
  }
  console.warn(
    `ALERT: ${domain} is ${result.severity}, ${result.daysRemaining} day(s) remaining, statuses=${JSON.stringify(statuses)}. ` +
    "Renew at the registrar, this cannot be fixed through the DNS provider API."
  );
}

export async function run() {
  console.log(`Checking domain expiration for ${DNS_DOMAIN} (DRY_RUN=${DRY_RUN})`);

  const expirationIso = await fetchRdapExpiration(DNS_DOMAIN);
  const statuses = await fetchRdapStatus(DNS_DOMAIN);
  const nowIso = new Date().toISOString();

  const result = daysUntilExpiry(expirationIso, nowIso, WARN_THRESHOLDS_DAYS);
  console.log(
    `${DNS_DOMAIN} expires ${expirationIso}: ${result.daysRemaining} day(s) remaining, severity=${result.severity}`
  );

  const graceFlags = new Set(["pendingdelete", "redemptionperiod", "autorenewperiod"]);
  const loweredStatuses = statuses.map((s) => s.toLowerCase());
  if (loweredStatuses.some((s) => graceFlags.has(s))) {
    console.warn(`${DNS_DOMAIN} is already in a grace or pending-delete state: ${JSON.stringify(statuses)}`);
    sendAlert(DNS_DOMAIN, result, statuses);
    return;
  }

  if (["warning", "critical", "expired"].includes(result.severity)) {
    sendAlert(DNS_DOMAIN, result, statuses);
  } else {
    console.log(`OK: ${DNS_DOMAIN} has plenty of runway left before renewal is needed.`);
  }
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  run().catch((e) => { console.error(e); process.exit(1); });
}
